File: src/cpp/server/MovieDescription.cpp
/**
Copyright [2016] [Eric Kaufman]

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributd on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 **/


#include "MovieDescription.hpp"
#include <jsonrpccpp/server.h>



MovieDescription::MovieDescription(string title, string rated, string released, string runtime, string plot, string filename, vector<string> genre, vector<string> actors)
:   title(title),
    rated(rated),
    released(released),
    runtime(runtime),
    plot(plot),
    filename(filename),
    genre(genre),
    actors(actors)
{
}

MovieDescription::~MovieDescription()
{
}

    string MovieDescription::getTitle(){
        return title;
    }

    string MovieDescription::getRated(){
        return rated;
    }
    
    string MovieDescription::getReleased(){
        return released;
    }
    
    string MovieDescription::getRuntime(){
        return runtime;
    }
    
    string MovieDescription::getPlot(){
        return plot;
    }   
    
    string MovieDescription::getFilename(){
        return filename;
    }
    
    vector<string> MovieDescription::getGenre(){
        return genre;
    }
    
    vector<string> MovieDescription::getActors(){
        return actors;
    }
    
    void MovieDescription::setTitle(string _title){
        title = _title;
    }
    void MovieDescription::setRated(string _rated){
        rated = _rated;
    }
    void MovieDescription::setReleased(string _released){
        released = _released;
    }
    void MovieDescription::setRuntime(string _runtime){
        runtime = _runtime;
    }
    void MovieDescription::setPlot(string _plot){
        plot = _plot;
    }
    void MovieDescription::setFilename(string _filename){
        filename = _filename;
    }
    void MovieDescription::setGenre(vector<string> _genre){
        genre = _genre;
    }
    void MovieDescription::setActors(vector<string> _actors){
        actors = _actors;
    }
    
    string MovieDescription::toString(){
        string genreString;
        string actorsString;
        
        for(size_t i = 0; i < genre.size(); i++){
            genreString += genre[i];
        }
        
        for(size_t i = 0; i < actors.size(); i++){
            actorsString += actors[i];
        }
        
        string s = "Title: " + title +
                "\n Rated: " + rated +
                "\n Release Date: " + released +
                "\n Runtime: " + runtime +
                "\n Plot: " + plot +
                "\n Filename: " + filename +
                "\n Genre: " + genreString +
                "\n Actors: " + actorsString;

        return s;
    }
	
	Json::Value MovieDescription::toJson(){
		Json::Value jsonObj;
		Json::Value jsonBody;
		jsonBody["Title"] = title;
		jsonBody["Rated"] = rated;
		jsonBody["Released"] = released;
		jsonBody["Runtime"] = runtime;
		jsonBody["Plot"] = plot;
		jsonBody["Filename"] = filename;
		Json::Value jsonGenre (Json::arrayValue);
		for(size_t i = 0; i < genre.size(); i++){
			jsonGenre.append(genre[i]);
		}
		
		Json::Value jsonActors (Json::arrayValue);
		for(size_t i = 0; i < actors.size(); i++){
			jsonActors.append(actors[i]);
		}
		jsonBody["Genre"] = jsonGenre;
		jsonBody["Actors"] = jsonActors;
		
		jsonObj[title] = jsonBody;
		return jsonObj;
		
	}

File: src/java/client/MovieDescription.java
import java.util.Vector;
import org.json.*;


/**Copyright [2016] [Eric Kaufman]

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 * Created by Eric on 8/29/2016.
 */

public class MovieDescription {

    private String title;
    private String rated;
    private String released;
    private String runtime;
    private String plot;
    private String filename;
    private Vector<String> genre;
    private Vector<String> actors;

    /**
     * Constructs a new MovieDescription object
     * @param title Title of movie
     * @param rated Rating of movie
     * @param released Release date of movie
     * @param runtime Length of movie in minutes
     * @param plot Plot of movie
     * @param filename Filename of movie file
     * @param genre List of applicable genres of movie
     * @param actors List of actors in movie
     */

    public MovieDescription(String title, String rated, String released, String runtime, String plot, String filename, Vector<String> genre, Vector<String> actors) {
        this.title = title;
        this.rated = rated;
        this.released = released;
        this.runtime = runtime;
        this.plot = plot;
        this.filename = filename;
        this.genre = genre;
        this.actors = actors;
    }

    public MovieDescription(org.json.JSONObject jsonObj){
        title = jsonObj.getString("Title");
        rated = jsonObj.getString("Rated");
        released = jsonObj.getString("Released");
        runtime = jsonObj.getString("Runtime");
        plot = jsonObj.getString("Plot");
        filename = jsonObj.getString("Filename");
        JSONArray jsonArr = jsonObj.getJSONArray("Genre");
        genre = new Vector<>();
        actors = new Vector<>();
        for(int i = 0; i < jsonArr.length(); i++){
            genre.add(jsonArr.getString(i));
        }
        jsonArr = jsonObj.getJSONArray("Actors");
        for(int i = 0; i < jsonArr.length(); i++){
            actors.add(jsonArr.getString(i));
        }

    }

    public  JSONObject toJson(){
        JSONObject jsonObj = new JSONObject();
        jsonObj.put("Title", title);
        jsonObj.put("Rated", rated);
        jsonObj.put("Released", released);
        jsonObj.put("Runtime", runtime);
        jsonObj.put("Plot", plot);
        jsonObj.put("Filename", filename);
        jsonObj.put("Genre", genre);
        jsonObj.put("Actors", actors);
        return jsonObj;
    }

    /**
     * Returns a human readable string description of this MovieDescription object
     * @return String containing the fields of the MovieDescription object
     */
    public String toString() {
        String s = "Title: " + title +
                "\n Rated: " + rated +
                "\n Release Date: " + released +
                "\n Runtime: " + runtime +
                "\n Plot: " + plot +
                "\n Filename: " + filename +
                "\n Genre: " + genre +
                "\n Actors: " + actors;

        return s;
    }

    //The rest of the class consists of getter and setter methods.  I'm not going to write comments on them as there are
    //quite a lot and their functionality is self-explanatory.

    public String getTitle() {
        return title;
    }

    public String getRated() {
        return rated;
    }

    public String getReleased() {
        return released;
    }

    public String getRuntime() {
        return runtime;
    }

    public String getPlot() {
        return plot;
    }

    public String getFilename() {
        return filename;
    }

    public Vector<String> getGenre() {
        return genre;
    }

    public Vector<String> getActors() {
        return actors;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public void setRated(String rated) {
        this.rated = rated;
    }

    public void setReleased(String released) {
        this.released = released;
    }

    public void setRuntime(String runtime) {
        this.runtime = runtime;
    }

    public void setPlot(String plot) {
        this.plot = plot;
    }

    public void setFilename(String filename) {
        this.filename = filename;
    }

    public void setGenre(Vector<String> genre) {
        this.genre = genre;
    }

    public void setActors(Vector<String> actors) {
        this.actors = actors;
    }

}
